packages/@aws-cdk/aws-cloudfront/lib/edge-function.ts

```diff
@@ -0,0 +1,271 @@
+      runtime: CustomResourceProviderRuntime.NODEJS_12,
+      policyStatements: [{
+        Effect: 'Allow',
+        Resource: parameterArnPrefix,
+        Action: ['ssm:GetParameter'],
+      }],
+    });
+    const resource = new CustomResource(this, 'ArnReader', {
+      resourceType: resourceType,
+      serviceToken,
+      properties: {
+        Region: EdgeFunctionExperimental.EDGE_REGION,
+        ParameterName: parameterName,
+        RefreshToken: edgeFunctionHash,
+      },
+    });
+    const edgeArn = resource.getAttString('FunctionArn');
+
+    return { edgeFunction, edgeArn, functionStack };
+  }
+
+  private edgeStack(): CrossRegionLambdaStack {
+    const stage = this.node.root;
+    if (!stage || !Stage.isStage(stage)) {
+      throw new Error('stacks which use EdgeFunctions must be part of a CDK app or stage');
+    }
+    const region = this.env.region;
+    if (Token.isUnresolved(region)) {
+      throw new Error('stacks which use EdgeFunctions must have an explicitly set region');
+    }
+
+    const edgeStackId = `edge-lambda-stack-${region}`;
+    let edgeStack = stage.node.tryFindChild(edgeStackId) as CrossRegionLambdaStack;
+    if (!edgeStack) {
+      edgeStack = new CrossRegionLambdaStack(stage, edgeStackId, {
+        synthesizer: this.crossRegionSupportSynthesizer(),
+        env: { region: EdgeFunctionExperimental.EDGE_REGION },
+      });
+    }
+    return edgeStack;
+  }
+
+  // Stolen from `@aws-cdk/aws-codepipeline`'s `Pipeline`.
+  private crossRegionSupportSynthesizer(): IStackSynthesizer | undefined {
+    // If we have the new synthesizer we need a bootstrapless copy of it,
+    // because we don't want to require bootstrapping the environment
+    // of the account in this replication region.
+    // Otherwise, return undefined to use the default.
+    return (this.stack.synthesizer instanceof DefaultStackSynthesizer)
+      ? new BootstraplessSynthesizer({
+        deployRoleArn: this.stack.synthesizer.deployRoleArn,
+        cloudFormationExecutionRoleArn: this.stack.synthesizer.cloudFormationExecutionRoleArn,
+      })
+      : undefined;
+  }
+}
+
+/** Result of creating an in-region or cross-region function */
+interface FunctionConfig {
+  readonly edgeFunction: lambda.Function;
+  readonly edgeArn: string;
+  readonly functionStack: Stack;
+}
+
+class CrossRegionLambdaStack extends Stack {
+
+  constructor(scope: CoreConstruct, id: string, props: StackProps) {
+    super(scope, id, props);
+  }
+
+  public addEdgeFunction(id: string, parameterName: string, props: lambda.FunctionProps) {
+    const role = props.role ?? defaultLambdaRole(this, id);
+
+    const edgeFunction = new lambda.Function(this, id, {
+      ...props,
+      role,
+    });
+
+    new ssm.StringParameter(edgeFunction, 'Parameter', {
+      parameterName,
+      stringValue: edgeFunction.currentVersion.edgeArn,
+    });
+
+    return edgeFunction;
+  }
+}
+
+function defaultLambdaRole(scope: Construct, id: string): iam.IRole {
```

Comment:
Would it make more sense to simply mutate the role created implicitly by `lambda.Function` instead? Otherwise, any changes to that role will have to be copy & pasted here.